applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_base_element.cpp

```diff
@@ -0,0 +1,603 @@
+// KRATOS  ___|  |                   |                   |
+//       \___ \  __|  __| |   |  __| __| |   |  __| _` | |
+//             | |   |    |   | (    |   |   | |   (   | |
+//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
+//
+//  License:		 BSD License
+//					 license: structural_mechanics_application/license.txt
+//
+//  Main authors:    Armin Geiser, https://github.com/armingeiser
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "adjoint_finite_difference_base_element.h"
+#include "structural_mechanics_application_variables.h"
+#include "custom_response_functions/response_utilities/response_data.h"
+#include "includes/checks.h"
+
+
+namespace Kratos
+{
+
+AdjointFiniteDifferencingBaseElement::AdjointFiniteDifferencingBaseElement(Element::Pointer pPrimalElement)
+                    : Element(pPrimalElement->Id(), pPrimalElement->pGetGeometry(), pPrimalElement->pGetProperties()) {
+
+    mpPrimalElement = pPrimalElement;
+}
+
+AdjointFiniteDifferencingBaseElement::~AdjointFiniteDifferencingBaseElement() {}
+
+Element::Pointer AdjointFiniteDifferencingBaseElement::Create(Element::Pointer pPrimalElement) const
+{
+    return Kratos::make_shared<AdjointFiniteDifferencingBaseElement>(pPrimalElement);
+}
+
+void AdjointFiniteDifferencingBaseElement::EquationIdVector(EquationIdVectorType& rResult,
+    ProcessInfo& rCurrentProcessInfo)
+{
+    const SizeType num_dofs = 6 * GetGeometry().PointsNumber(); // 6 dofs per node
+    if(rResult.size() != num_dofs)
+        rResult.resize(num_dofs, false);
+
+    GeometryType & geom = this->GetGeometry();
+
+    for(SizeType i = 0; i < geom.size(); i++)
+    {
+        int index = i * 6;
+        NodeType & iNode = geom[i];
+
+        rResult[index]     = iNode.GetDof(ADJOINT_DISPLACEMENT_X).EquationId();
+        rResult[index + 1] = iNode.GetDof(ADJOINT_DISPLACEMENT_Y).EquationId();
+        rResult[index + 2] = iNode.GetDof(ADJOINT_DISPLACEMENT_Z).EquationId();
+
+        rResult[index + 3] = iNode.GetDof(ADJOINT_ROTATION_X).EquationId();
```

Comment:
Very good point!

I think @KlausBSautter is using sth for this in the Explicit Stuff, but I am not sure if this is the best solution

We should discuss it

Reply:
I agree, the rotation dofs should be optional. It needs to be discussed if this is done via an if/else statement in this class or using inheritance.

Reply:
However i suggest doing this in a separate PR, because we need an additional element that does not have rotation dofs to test it, and it would further increase this PR.